File: javascript/vehicle-purchase/vehicle-purchase.js
// @ts-check

const LICENSE_VEHICLES = ["car", "truck"];

/**
 * Determines whether or not you need a licence to operate a certain kind of vehicle.
 *
 * @param {string} kind
 * @returns {boolean} whether a license is required
 */
export function needsLicense(kind) {
  return LICENSE_VEHICLES.includes(kind.toLowerCase());
}

/**
 * Helps choosing between two options by recommending the one that
 * comes first in dictionary order.
 *
 * @param {string} option1
 * @param {string} option2
 * @returns {string} a sentence of advice which option to choose
 */
export function chooseVehicle(option1, option2) {

  let vehicle = undefined;

  if (option1 < option2) {
    vehicle = option1;
  } else {
    vehicle = option2;
  }

  return `${vehicle} is clearly the better choice.`;
}

/**
 * Calculates an estimate for the price of a used vehicle in the dealership
 * based on the original price and the age of the vehicle.
 *
 * @param {number} originalPrice
 * @param {number} age
 * @returns expected resell price in the dealership
 */
export function calculateResellPrice(originalPrice, age) {

  let discount = undefined;

  if (age < 3) {
    discount = 0.8;
  } else if (age <= 10) {
    discount = 0.7;
  } else {
    discount = 0.5;
  }

  return originalPrice * discount;
}

File: go/vehicle-purchase/vehicle_purchase.go
package purchase

var licensed = []string{"car", "truck"}

// NeedsLicense determines whether a license is needed to drive a type of vehicle. Only "car" and "truck" require a license.
func NeedsLicense(kind string) bool {

	// Go does not have 'contains' or 'in' etc. for array/slice membership
	for _, val := range licensed {

		if val == kind {
			return true
		}
	}

	return false
}

// ChooseVehicle recommends a vehicle for selection. It always recommends the vehicle that comes first in dictionary order.
func ChooseVehicle(option1, option2 string) string {

	better := option1

	if option2 < option1 {
		better = option2
	}

	return better + " is clearly the better choice."
}

// CalculateResellPrice calculates how much a vehicle can resell for at a certain age.
func CalculateResellPrice(originalPrice, age float64) float64 {

	var percent float64

	switch {
	case age < 3:
		percent = 0.8
	case age < 10:
		percent = 0.7
	default:
		percent = 0.5
	}

	return originalPrice * percent
}
